File: samples/java/src/main/java/com/alibaba/cloud/cms/samples/examples/ChatFromFile.java
package com.alibaba.cloud.cms.samples.examples;

import com.alibaba.cloud.cms.samples.client.*;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.*;
import java.nio.file.*;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.BlockingQueue;

/**
 * 从文件加载请求示例
 * Load requests from file example
 *
 * Usage:
 *   mvn exec:java -Dexec.mainClass="com.alibaba.cloud.cms.samples.examples.ChatFromFile" -Dexec.args="-file ../../requests/cms/entity.json"
 *   mvn exec:java -Dexec.mainClass="com.alibaba.cloud.cms.samples.examples.ChatFromFile" -Dexec.args="-dir ../../requests/cms/"
 */
public class ChatFromFile {
    private static String filePath = null;
    private static String dirPath = null;
    private static boolean simpleMode = false;
    private static String outputDir = "../../requests/output";
    private static final ObjectMapper objectMapper = new ObjectMapper();

    public static void main(String[] args) {
        parseArgs(args);

        if (filePath == null && dirPath == null) {
            printUsage();
            System.exit(1);
        }

        System.out.println("🚀 VibeOps Chat - 从文件加载请求 (Java)");
        System.out.println("=".repeat(60));

        try {
            // Load configuration
            Config cfg = Config.loadFromEnv();

            // Create client
            AgentClient client = new AgentClient(cfg);

            // Ensure output directory exists
            Files.createDirectories(Paths.get(outputDir));

            // Process files
            if (dirPath != null) {
                processDirectory(client, dirPath);
            } else {
                processFile(client, filePath);
            }

            client.shutdown();
        } catch (SDKException e) {
            System.out.printf("❌ 配置加载失败: %s%n", e.getMessage());
            System.exit(1);
        } catch (Exception e) {
            System.out.printf("❌ 错误: %s%n", e.getMessage());
            System.exit(1);
        }
    }

    private static void parseArgs(String[] args) {
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "-file":
                    if (i + 1 < args.length) filePath = args[++i];
                    break;
                case "-dir":
                    if (i + 1 < args.length) dirPath = args[++i];
                    break;
                case "-simple":
                    simpleMode = true;
                    break;
                case "-output":
                    if (i + 1 < args.length) outputDir = args[++i];
                    break;
            }
        }
    }

    private static void processDirectory(AgentClient client, String dir) throws IOException {
        File[] files = new File(dir).listFiles((d, name) -> name.endsWith(".json"));
        if (files == null || files.length == 0) {
            System.out.printf("⚠️ 目录中没有 JSON 文件: %s%n", dir);
            return;
        }

        System.out.printf("📁 找到 %d 个请求文件%n%n", files.length);

        for (int i = 0; i < files.length; i++) {
            System.out.printf("━━━ [%d/%d] %s ━━━%n", i + 1, files.length, files[i].getName());
            processFile(client, files[i].getPath());
            System.out.println();
        }

        System.out.printf("✅ 处理完成，共 %d 个文件%n", files.length);
    }

    private static void processFile(AgentClient client, String file) {
        try {
            // Load request file
            JsonNode reqFile = objectMapper.readTree(new File(file));

            // Extract message
            String message = extractMessage(reqFile);
            if (message == null || message.isEmpty()) {
                System.out.println("⚠️ 文件中没有消息内容");
                return;
            }

            System.out.printf("📄 文件: %s%n", new File(file).getName());
            System.out.printf("💬 消息: %s%n", truncate(message, 60));

            // Create thread
            String threadId = client.createThread();

            // Create output file
            PrintWriter outputFile = createOutputFile(file);

            // Write request info
            writeOutput(outputFile, "# Request: " + new File(file).getName());
            writeOutput(outputFile, "# Time: " + Instant.now().toString());
            writeOutput(outputFile, "# ThreadID: " + threadId);
            writeOutput(outputFile, "# Message: " + message + "\n");

            System.out.println("-".repeat(60));

            // Extract variables
            Map<String, Object> variables = extractVariables(reqFile);

            // Send request
            long startTime = System.currentTimeMillis();
            BlockingQueue<ChatEvent> events = client.chatWithVariables(threadId, message, variables);

            // Process response
            SimplePrinter printer = new SimplePrinter();
            int eventIndex = 0;

            while (true) {
                ChatEvent event = events.take();
                eventIndex++;

                if (event.hasError()) {
                    Exception err = event.getError();
                    System.out.printf("❌ 错误: %s%n", err.toString());
                    if (err.getCause() != null) {
                        System.out.printf("   原因: %s%n", err.getCause().getMessage());
                    }
                    writeOutput(outputFile, "[ERROR] " + err.toString());
                    break;
                }

                // Write raw event
                if (event.getRawJson() != null && !event.getRawJson().isEmpty()) {
                    writeOutput(outputFile, "[EVENT " + eventIndex + "]\n" + event.getRawJson() + "\n");
                }

                // Output
                String text = printer.processEvent(event);
                if (!text.isEmpty()) {
                    System.out.print(text);
                }

                if (event.isDone()) {
                    break;
                }
            }

            long elapsed = System.currentTimeMillis() - startTime;
            System.out.println();

            // Write final result
            if (simpleMode) {
                String finalText = printer.getFinalText();
                writeOutput(outputFile, "\n# Final Result:\n" + finalText);
                System.out.printf("📄 最终文本:%n%s%n", finalText);
            }

            writeOutput(outputFile, "\n# Duration: " + elapsed + "ms");
            System.out.printf("⏱️  耗时: %dms%n", elapsed);

            if (outputFile != null) {
                outputFile.close();
            }
        } catch (Exception e) {
            System.out.printf("❌ 处理文件失败: %s%n", e.getMessage());
        }
    }

    private static String extractMessage(JsonNode reqFile) {
        if (reqFile.has("messages") && reqFile.get("messages").isArray()) {
            JsonNode messages = reqFile.get("messages");
            if (messages.size() > 0) {
                JsonNode firstMsg = messages.get(0);
                if (firstMsg.has("contents") && firstMsg.get("contents").isArray()) {
                    JsonNode contents = firstMsg.get("contents");
                    if (contents.size() > 0 && contents.get(0).has("value")) {
                        return contents.get(0).get("value").asText();
                    }
                }
            }
        }
        return null;
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> extractVariables(JsonNode reqFile) {
        if (reqFile.has("variables")) {
            try {
                return objectMapper.convertValue(reqFile.get("variables"), Map.class);
            } catch (Exception ignored) {}
        }
        return new HashMap<>();
    }

    private static PrintWriter createOutputFile(String inputFile) {
        try {
            String baseName = new File(inputFile).getName().replace(".json", "");
            String timestamp = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd-HHmmss"));
            String outputPath = outputDir + "/" + baseName + "-" + timestamp + ".log";
            return new PrintWriter(new FileWriter(outputPath));
        } catch (IOException e) {
            return null;
        }
    }

    private static void writeOutput(PrintWriter file, String content) {
        if (file != null) {
            file.println(content);
        }
    }

    private static String truncate(String s, int maxLen) {
        if (s.length() <= maxLen) return s;
        return s.substring(0, maxLen) + "...";
    }

    private static void printUsage() {
        System.out.println("用法:");
        System.out.println("  -file <path>   处理单个文件");
        System.out.println("  -dir <path>    处理目录下所有 JSON 文件");
        System.out.println();
        System.out.println("示例:");
        System.out.println("  -file ../../requests/cms/entity.json");
        System.out.println("  -dir ../../requests/cms/");
        System.out.println("  -file entity.json -simple");
        System.out.println();
        System.out.println("选项:");
        System.out.println("  -file     请求 JSON 文件路径");
        System.out.println("  -dir      请求文件目录");
        System.out.println("  -simple   简洁模式，只输出最终文本");
        System.out.println("  -output   输出目录 (默认: output)");
    }
}

File: samples/python/cms_sdk_samples/examples/chat_from_file.py
#!/usr/bin/env python3
"""
从文件加载请求示例
Load requests from file example

Usage:
    python -m cms_sdk_samples.examples.chat_from_file -file ../../requests/cms/entity.json
    python -m cms_sdk_samples.examples.chat_from_file -dir ../../requests/cms/
"""

import argparse
import asyncio
import json
import os
import sys
import time
from datetime import datetime
from pathlib import Path

from ..client import AgentClient, Config, SDKException, SimplePrinter, EventPrinter


def parse_args():
    parser = argparse.ArgumentParser(description="从文件加载请求")
    parser.add_argument("-file", dest="file_path", help="请求 JSON 文件路径")
    parser.add_argument("-dir", dest="dir_path", help="请求文件目录")
    parser.add_argument("-simple", action="store_true", help="简洁模式")
    parser.add_argument("-output", default="../../requests/output", help="输出目录")
    return parser.parse_args()


async def process_file(client: AgentClient, file_path: str, output_dir: str, simple_mode: bool):
    try:
        # Load request file
        with open(file_path, "r", encoding="utf-8") as f:
            req_file = json.load(f)

        # Extract message
        message = extract_message(req_file)
        if not message:
            print("⚠️ 文件中没有消息内容")
            return

        file_name = Path(file_path).name
        print(f"📄 文件: {file_name}")
        print(f"💬 消息: {truncate(message, 60)}")

        # Create thread
        thread_id = client.create_thread()

        # Create output file
        output_file = create_output_file(file_path, output_dir)

        # Write request info
        write_output(output_file, f"# Request: {file_name}")
        write_output(output_file, f"# Time: {datetime.now().isoformat()}")
        write_output(output_file, f"# ThreadID: {thread_id}")
        write_output(output_file, f"# Message: {message}\n")

        print("-" * 60)

        # Extract variables
        variables = req_file.get("variables", {})

        # Send request
        start_time = time.time()
        simple_printer = SimplePrinter() if simple_mode else None
        event_printer = None if simple_mode else EventPrinter(print_raw_body=False, print_parsed=True)
        event_index = 0

        async for event in client.chat_with_variables(thread_id, message, variables):
            event_index += 1

            if event.has_error():
                print(f"❌ 错误: {event.error}")
                write_output(output_file, f"[ERROR] {event.error}")
                break

            # Write raw event
            if event.raw_json:
                write_output(output_file, f"[EVENT {event_index}]\n{event.raw_json}\n")

            # Output
            if simple_mode:
                text = simple_printer.process_event(event)
                if text:
                    print(text, end="", flush=True)
            else:
                event_printer.print_event(event, event_index)

        elapsed = time.time() - start_time
        print()

        # Write final result
        if simple_mode and simple_printer:
            final_text = simple_printer.get_final_text()
            write_output(output_file, f"\n# Final Result:\n{final_text}")
            print(f"📄 最终文本:\n{final_text}")

        write_output(output_file, f"\n# Duration: {elapsed:.2f}s")
        print(f"⏱️  耗时: {elapsed:.2f}s")

        if output_file:
            output_file.close()
            print(f"📁 输出: {output_file.name}")

    except Exception as e:
        print(f"❌ 处理文件失败: {e}")


async def process_directory(client: AgentClient, dir_path: str, output_dir: str, simple_mode: bool):
    files = list(Path(dir_path).glob("*.json"))
    if not files:
        print(f"⚠️ 目录中没有 JSON 文件: {dir_path}")
        return

    print(f"📁 找到 {len(files)} 个请求文件\n")

    for i, file_path in enumerate(files):
        print(f"━━━ [{i + 1}/{len(files)}] {file_path.name} ━━━")
        await process_file(client, str(file_path), output_dir, simple_mode)
        print()

    print(f"✅ 处理完成，共 {len(files)} 个文件")


def extract_message(req_file: dict) -> str:
    messages = req_file.get("messages", [])
    if messages and messages[0].get("contents"):
        contents = messages[0]["contents"]
        if contents and contents[0].get("value"):
            return contents[0]["value"]
    return ""


def create_output_file(input_file: str, output_dir: str):
    try:
        os.makedirs(output_dir, exist_ok=True)
        base_name = Path(input_file).stem
        timestamp = datetime.now().strftime("%Y%m%d-%H%M%S")
        output_path = os.path.join(output_dir, f"{base_name}-{timestamp}.log")
        return open(output_path, "w", encoding="utf-8")
    except Exception:
        return None


def write_output(f, content: str):
    if f:
        f.write(content + "\n")


def truncate(s: str, max_len: int) -> str:
    if len(s) <= max_len:
        return s
    return s[:max_len] + "..."


async def main_async():
    args = parse_args()

    if not args.file_path and not args.dir_path:
        print("用法:")
        print("  -file <path>   处理单个文件")
        print("  -dir <path>    处理目录下所有 JSON 文件")
        sys.exit(1)

    print("🚀 VibeOps Chat - 从文件加载请求 (Python)")
    print("=" * 60)

    try:
        # Load configuration
        cfg = Config.load_from_env()

        # Create client
        client = AgentClient(cfg)

        # Ensure output directory exists
        os.makedirs(args.output, exist_ok=True)

        # Process files
        if args.dir_path:
            await process_directory(client, args.dir_path, args.output, args.simple)
        else:
            await process_file(client, args.file_path, args.output, args.simple)

    except SDKException as e:
        print(f"❌ 配置加载失败: {e}")
        sys.exit(1)
    except Exception as e:
        print(f"❌ 错误: {e}")
        sys.exit(1)


def main():
    asyncio.run(main_async())


if __name__ == "__main__":
    main()
